src/sync/stream.ts

```diff
@@ -61,6 +74,147 @@ function isStreamSyncHistoryError(data: IStreamSyncHistoryData): data is IStream
   return data.type === 'stream:history:error';
 }
 
+interface IStreamItemVertex {
+  seq: number;
+  type: 'vertex';
+  vertex: IHistoryTx;
+}
+
+interface IStreamItemAddress {
+  seq: number;
+  type: 'address';
+  address: IAddressInfo;
+}
+
+type IStreamItem = IStreamItemAddress | IStreamItemVertex;
+
+function isStreamItemVertex(item: IStreamItem): item is IStreamItemVertex {
+  return item.type === 'vertex';
+}
+
+function isStreamItemAddress(item: IStreamItem): item is IStreamItemAddress {
+  return item.type === 'address';
+}
+
+/**
+ * Stream statistics manager.
+ * Will provide insight on the rates of the stream and how the client is performing.
+ */
+class StreamStatsManager {
+  // Counter for received events
+  recvCounter: number;
+
+  // Counter for processed events
+  procCounter: number;
+
+  // Counter for ack messages sent
+  ackCounter: number;
+
+  // Counter for the number of times the queue has been empty.
+  emptyCounter: number;
+
+  // Timer to process in_rate, which is number of received events per second.
+  inTimer?: ReturnType<typeof setTimeout>;
+
+  // Timer to process out_rate, which is the number of processed events per second.
+  outTimer?: ReturnType<typeof setTimeout>;
+
+  // Simple timer to log the number of events on queue.
+  qTimer?: ReturnType<typeof setTimeout>;
+
+  // Custom logger to log in the desired format.
+  logger: ILogger;
+
+  // Reference to the item queue we are processing.
+  q: Queue;
+
+  // Interval to use when calculating rates.
+  dt: number;
+
+  constructor(queue: Queue, logger: ILogger) {
+    this.recvCounter = 0;
+    this.procCounter = 0;
+    this.ackCounter = 0;
+    this.emptyCounter = 0;
+
+    this.logger = logger;
+    this.dt = 2000;
```

Comment:
So this is the sample rate for the `in_rate` and `out_rate`, right?

Reply:
Yes

Reply:
Thanks for renaming it.